features/mbedtls/platform/src/threading_alt.cpp

```diff
@@ -0,0 +1,79 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_THREADING_C)
+
+#include "mbedtls/threading.h"
+
+#if defined(MBEDTLS_THREADING_ALT)
+static void threading_mutex_init_mbed( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL )
+        return;
+
+    mutex->mutex = new rtos::Mutex();
+
+    mutex->is_valid = true;
+}
+
+static void threading_mutex_free_mbed( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL || mutex->is_valid != true )
+        return;
+
+    delete mutex->mutex;
+
+    mutex->is_valid = false;
+}
+
+static int threading_mutex_lock_mbed( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL || mutex->is_valid != true )
+        return( MBEDTLS_ERR_THREADING_BAD_INPUT_DATA );
+
```

Comment:
I think that the only case `mutex->mutex` can be `NULL` is when 
- `mutex` is uninitialised OR 
- the memory is corrupted / the structure has been accessed with some means other than these functions

In these cases if we suppose that the memory content is uniformly distributed, then the chance of not catching this with `is_valid` is 1:2^8 and therefore detecting this by checking for `NULL` is 1:2^40 or 1:2^72, depending on the architecture (it can be higher on architectures with smaller address space). 

The check probably wouldn't help against an active adversary at all, because an active adversary would probably set the pointer value to something other than `NULL` if he can.

I think if we want to make it more robust then changing the type of `is_valid` to `int` would be much more effective, than adding this check for `NULL`